File: scala/src/main/scala/tddmicroexercises/braking/Brake.scala
package tddmicroexercises.braking


class Brake(var strategy: AbstractBrakingStrategy) {
  def this() = { this(new SteadyBrakingStrategy()) }

  def getPeakForce: Double = {
    return Brake.peakForce
  }

  def setPeakForce(peakForce: Double) {
    Brake.peakForce = peakForce
  }

  def powerDissipation(velocity: Double): Double = {
    return strategy.powerDissipation(velocity, this)
  }

  def forceAtBrakeLevel(percentOfPeakForce: Double): Double = {
    return Brake.peakForce * percentOfPeakForce
  }
}

object Brake {
  private var peakForce: Double = 30
}

File: Java/Braking/src/main/java/tddmicroexercises/braking/Brake.java
package tddmicroexercises.braking;

public class Brake {
    private BrakingStrategy strategy;
    private static double peakForce = 30;

    public Brake(BrakingStrategy strategy) {
        this.strategy = strategy;
    }

    public double getPeakForce() {
        return peakForce;
    }

    public void setPeakForce(double peakForce) {
        this.peakForce = peakForce;
    }

    public double powerDissipation(double velocity) {
        if (this.strategy == null) {
            if (velocity < 60) {
                this.strategy = new SteadyBrakingStrategy();
            } else if (velocity < 120) {
                this.strategy = new IntervalBrakingStrategy();
            } else {
                this.strategy = new FailingBrakingStrategy();
            }
        } 
        return strategy.powerDissipation(velocity, this);
    }

    public double forceAtBrakeLevel(double percentOfPeakForce) {
        return this.peakForce*percentOfPeakForce;
    }
}
